src/OpenTelemetry.Instrumentation.AspNetCore/Implementation/HttpInListener.cs

```diff
@@ -242,6 +244,15 @@ public void OnStopActivity(Activity activity, object payload)
 
                 activity.SetTag(SemanticConventions.AttributeHttpStatusCode, TelemetryHelper.GetBoxedStatusCode(response.StatusCode));
 
+#if NET6_0_OR_GREATER
+                var route = (context.GetEndpoint() as RouteEndpoint)?.RoutePattern.RawText;
+                if (!string.IsNullOrEmpty(route))
+                {
+                    activity.DisplayName = route;
```

Comment:
I've tried things out on an MVC app before and after this PR. I think we need to consider the test matrix a bit before going forward with this PR.

Below is abbreviated output from the console exporter highlighting the important differences. Three endpoints are exercised:

* http://localhost:5199/
* http://localhost:5199/Home/Index
* http://localhost:5199/Home/Privacy

Two things of note:

1. Note that the display name of the Activity is now the same for every endpoint.
2. Now that the `http.route` attribute is set, it too is the same for all of the endpoints - I'm not sure if this is a simply a limitation due to the type of routing I'm using, or if there's a way we can get the actual controller/action of the route in the `http.route`.

**Before**
Activity.DisplayName:        /
Activity.Tags:
    http.target: /
    http.url: http://localhost:5199/

Activity.DisplayName:        /Home/Index
Activity.Tags:
    http.target: /Home/Index
    http.url: http://localhost:5199/Home/Index

Activity.DisplayName:        /Home/Privacy
Activity.Tags:
    http.target: /Home/Privacy
    http.url: http://localhost:5199/Home/Privacy

**After**
Activity.DisplayName:        {controller=Home}/{action=Index}/{id?}
Activity.Tags:
    http.target: /
    http.url: http://localhost:5199/
    http.route: {controller=Home}/{action=Index}/{id?}

Activity.DisplayName:        {controller=Home}/{action=Index}/{id?}
Activity.Tags:
    http.target: /Home/Index
    http.url: http://localhost:5199/Home/Index
    http.route: {controller=Home}/{action=Index}/{id?}

Activity.DisplayName:        {controller=Home}/{action=Index}/{id?}
Activity.Tags:
    http.target: /Home/Privacy
    http.url: http://localhost:5199/Home/Privacy
    http.route: {controller=Home}/{action=Index}/{id?}

Reply:
Just a quick brainstorm of the various routing use cases we might need coverage for.

* [Conventional routing](https://learn.microsoft.com/en-us/aspnet/core/mvc/controllers/routing?view=aspnetcore-7.0#conventional-routing)
* [Attribute-based routing](https://learn.microsoft.com/en-us/aspnet/core/mvc/controllers/routing?view=aspnetcore-7.0#attribute-routing-for-rest-apis)
* [Minimal API](https://learn.microsoft.com/en-us/aspnet/core/fundamentals/routing?view=aspnetcore-7.0#endpoints)
* [Verb templates](https://learn.microsoft.com/en-us/aspnet/core/mvc/controllers/routing?view=aspnetcore-7.0#http-verb-templates)
* There may be other things that may or may not be relevant to us. Like [Areas](https://learn.microsoft.com/en-us/aspnet/core/mvc/controllers/routing?view=aspnetcore-7.0#areas)

I'm not suggesting we need coverage for all these things in this PR, but given that routing is so complex, I think it'd be nice to be able to have a table to be able to point to that makes it clear what to expect when using different styles of routing. Something like:


| | Activity.DisplayName | Span http.route | Metric http.route |
| ---- | ---- | ---- | ---- |
| Conventional routing | ... | ... | ... |
| Attribute-based routing | ... | ... | ... |
| ... | ... | ... | ...|

Reply:
@alanwest - Have updated PR description. I think the case you had https://github.com/open-telemetry/opentelemetry-dotnet/pull/4104#discussion_r1194399908 is [conventional routing](https://learn.microsoft.com/en-us/aspnet/core/mvc/controllers/routing?view=aspnetcore-7.0#set-up-conventional-route)

We could have these added to test

Reply:
With this PR if `/Home/Privacy` maps to `{controller=Home}/{action=Index}/{id?}` with conventional routing, then I think we should resolve this before merging.

Reply:
It looks like you have expected behavior.

I haven't run your tests, but I guess you're getting one route for http://localhost:5199/ and http://localhost:5199/Home/Index and http://localhost:5199/Home/Privacy because they all match to the same conventional route. That's how conventional routing works: an app centrally specifies just a few route patterns, then those patterns match to multiple controllers and actions.

The pattern `{controller=Home}/{action=Index}/{id?}` will take any URL with 0-3 segments, e.g. `/`, `/Home/Index`, `/Home/Privacy`, and match the segments up to the conventional route.

You could add some customization like replacing `controller`, `action` and `page` parameters with values from `RouteData`. That might be more useful, but it wouldn't be the original template anymore.

Reply:
Thanks @JamesNK for reviewing this. We have a similar issue with API versioning https://github.com/open-telemetry/opentelemetry-dotnet/issues/4525. Where the template has same value for different API versions. Would that one also require some customization?